File: go/test/driver_test.go
/* The DRIVER sections: everything a command list drives. */

package plugintest

import "testing"

var driverSections = []string{
	"lifecycle", "order", "point", "export", "depend",
	"declare", "state", "resource", "nest", "trace", "apply", "error",
}

func TestDriverSections(t *testing.T) {
	for _, name := range driverSections {
		groups, err := Section(name)
		if nil != err {
			t.Fatalf("%s: %v", name, err)
		}

		t.Run(name+": every entry carries cmd", func(t *testing.T) {
			for _, g := range Groups(groups) {
				for i, e := range groups[g] {
					if nil == e.Cmd {
						t.Errorf("driver entry without cmd: %s", Label(g, i, e))
					}
				}
			}
		})

		for _, g := range Groups(groups) {
			name, g := name, g
			t.Run(name+"/"+g, func(t *testing.T) {
				for i, e := range groups[g] {
					why := Check(e, func(en Entry) (any, error) { return Drive(en.Cmd) })
					if "" != why {
						t.Errorf("%s: %s", Label(g, i, e), why)
					}
				}
			})
		}
	}
}

File: javascript/test/driver.test.js
/* The DRIVER sections: everything a command list drives. */

'use strict'

const { test } = require('node:test')
const Assert = require('node:assert')

const { section, check, label } = require('./corpus')
const { drive } = require('./driver')

const DRIVER_SECTIONS = [
  'lifecycle', 'order', 'point', 'export', 'depend',
  'declare', 'state', 'resource', 'nest', 'trace', 'apply', 'error',
]

for (const name of DRIVER_SECTIONS) {
  const groups = section(name)

  test(name + ': every entry carries a command list in `in`', () => {
    const bad = []
    for (const g of Object.keys(groups)) {
      groups[g].forEach((e, i) => {
        if (!Array.isArray(e.in)) bad.push(label(g, i, e))
      })
    }
    Assert.deepEqual(bad, [], 'driver entries without a command list in `in`: ' + bad.join(', '))
  })

  for (const g of Object.keys(groups)) {
    test(name + '/' + g, () => {
      const fails = []
      groups[g].forEach((e, i) => {
        const why = check(e, (en) => drive(en.in))
        if (why) fails.push(label(g, i, e) + ': ' + why)
      })
      Assert.deepEqual(fails, [], fails.join('\n'))
    })
  }
}

module.exports = { DRIVER_SECTIONS }
